### PR TITLE
drivers: crypto: Add support for STM32L4 AES accelerator

### PR DESCRIPTION
**Pull request by lucasdietrich**
The STM32L4 series features an AES accelerator with differences in the HAL API compared to other STM32 devices. This patch adds support for the STM32L4 AES accelerator in the STM32 crypto driver.

Key changes:

- Introduce a new device type `st,stm32l4-aes` to represent the STM32L4 AES accelerator variant. Add the device tree binding file `st,stm32l4-aes.yaml` and update `stm32l462.dtsi` accordingly.
- Modify `crypto_stm32.c` to handle the differences in the HAL API:
  - Use specific HAL functions for encryption and decryption (e.g., `HAL_CRYP_AESECB_Encrypt`) when building for STM32L4 AES.
  - Adjust the initialization and configuration structures accordingly.
  - Replace `copy_reverse_words` with `copy_words_adjust_endianness` to handle endianness conversion only when necessary.

The driver has been successfully tested with the `crypto` sample on a custom board based on the STM32L462RE.

---

Related to issue #78482

### CODE REVIEW

File: drivers/crypto/crypto_stm32.c [hunk 1]
```diff
@@ -159,24 +188,23 @@ static int crypto_stm32_ecb_decrypt(struct cipher_ctx *ctx,
...
 	if (ret == 0) {
 		pkt->out_len = 16;
 	}
 
 	return ret;
 }
 
-static int crypto_stm32_cbc_encrypt(struct cipher_ctx *ctx,
-				    struct cipher_pkt *pkt, uint8_t *iv)
+static int crypto_stm32_cbc_encrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt, uint8_t *iv)
```
Comment (erwango) [inline drivers/crypto/crypto_stm32.c:199]: Avoid style changes if not required, it confuses review w/o additional review.
If really required, make it a dedicated commit.

File: drivers/crypto/crypto_stm32.c [hunk 2]
```diff
@@ -75,17 +115,22 @@ static int do_encrypt(struct cipher_ctx *ctx, uint8_t *in_buf, int in_len,
...
+	 * called bellow. */
+	memcpy(&data->hcryp.Init, &session->config, sizeof(session->config));
+#endif
 
-	status = HAL_CRYP_Encrypt(&data->hcryp, (uint32_t *)in_buf, in_len,
-				  (uint32_t *)out_buf, HAL_MAX_DELAY);
+	status = fn(&data->hcryp, in_buf, in_len, out_buf, HAL_MAX_DELAY);
 	if (status != HAL_OK) {
-		LOG_ERR("Encryption error");
+		LOG_ERR("Encryption/decryption error");
```
Comment (erwango) [inline drivers/crypto/crypto_stm32.c:133]: Split overall changes added to the driver (which will impact already supported IPs) and L4 support additions in distinct commits.

File: dts/arm/st/l4/stm32l462.dtsi [hunk 3]
```diff
@@ -11,7 +11,7 @@
 		compatible = "st,stm32l462", "st,stm32l4", "simple-bus";
 
 		aes: aes@50060000 {
-			compatible = "st,stm32-aes";
+			compatible = "st,stm32l4-aes";
```
Comment (erwango) [inline dts/arm/st/l4/stm32l462.dtsi:14]: Note: Alternatively, you can do:
```suggestion
			compatible = "st,stm32-aes", "st,stm32l4-aes";
```
This way, you benefit from `DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)` being defined and it can simplify things like:
```diff
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp)
#define DT_DRV_COMPAT st_stm32_cryp
#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
#define DT_DRV_COMPAT st_stm32_aes
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)    ===> No more required
+#define DT_DRV_COMPAT st_stm32l4_aes
```

Comment (lucasdietrich) [reply dts/arm/st/l4/stm32l462.dtsi]: Consequently, I introduced USE_HAL_GENERIC and USE_HAL_L4 to differentiate the nature of the two HALs

File: drivers/crypto/crypto_stm32.c [hunk 4]
```diff
@@ -47,11 +47,31 @@ LOG_MODULE_REGISTER(crypto_stm32);
 
 struct crypto_stm32_session crypto_stm32_sessions[CRYPTO_MAX_SESSION];
 
-static int copy_reverse_words(uint8_t *dst_buf, int dst_len,
-			      const uint8_t *src_buf, int src_len)
-{
-	int i;
-
```
Comment (erwango) [inline drivers/crypto/crypto_stm32.c]: Something is broken in your commit split.

File: drivers/crypto/crypto_stm32.c [hunk 5]
```diff
@@ -45,7 +45,11 @@ LOG_MODULE_REGISTER(crypto_stm32);
 #define STM32_CRYPTO_TYPEDEF            AES_TypeDef
 #endif
 
-struct crypto_stm32_session crypto_stm32_sessions[CRYPTO_MAX_SESSION];
+#define USE_HAL_GENERIC                                                                            \
+	(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) ||                                               \
+	 (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes) && !DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)))
+
+#define USE_HAL_L4 (DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes))
```
Comment (erwango) [inline drivers/crypto/crypto_stm32.c:52]: Those 2 defines are not required. Checking `DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)` should be sufficient.

File: drivers/crypto/crypto_stm32.c [hunk 6]
```diff
@@ -95,35 +114,21 @@ static int do_encrypt(struct cipher_ctx *ctx, uint8_t *in_buf, int in_len,
 	return 0;
 }
 
-static int do_decrypt(struct cipher_ctx *ctx, uint8_t *in_buf, int in_len,
-		      uint8_t *out_buf)
+#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
```
Comment (erwango) [inline drivers/crypto/crypto_stm32.c]: Nitpick: `!DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)` is not actually required in this commit

File: drivers/crypto/crypto_stm32.c [hunk 7]
```diff
@@ -47,11 +47,31 @@ LOG_MODULE_REGISTER(crypto_stm32);
...
+ *                    the configuration information for the CRYP module.
+ * @param in_data     Pointer to input data (plaintext for encryption or ciphertext for decryption).
+ * @param size        Length of the input data in bytes.
+ * @param out_data    Pointer to output data (ciphertext for encryption or plaintext for
+ * decryption).
+ * @param timeout     Timeout duration in milliseconds.
+ *
+ * @retval HAL_StatusTypeDef  HAL status of the operation.
+ */
+typedef HAL_StatusTypeDef (*hal_cryp_aes_op_func_t)(CRYP_HandleTypeDef *hcryp, uint8_t *in_data,
```
Comment (erwango) [inline drivers/crypto/crypto_stm32.c:69]:
```suggestion
typedef status_t (*hal_cryp_aes_op_func_t)(CRYP_HandleTypeDef *hcryp, uint8_t *in_data,
```

File: drivers/crypto/crypto_stm32_priv.h [hunk 8]
```diff
@@ -8,6 +8,12 @@
 #ifndef ZEPHYR_DRIVERS_CRYPTO_CRYPTO_STM32_PRIV_H_
 #define ZEPHYR_DRIVERS_CRYPTO_CRYPTO_STM32_PRIV_H_
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
+#define STM32_CRYPTO_CONFIG_TYPEDEF            CRYP_InitTypeDef
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
```
Comment (erwango) [inline drivers/crypto/crypto_stm32_priv.h:13]: `#else` is sufficient here

File: drivers/crypto/crypto_stm32_priv.h [hunk 9]
```diff
@@ -8,6 +8,12 @@
 #ifndef ZEPHYR_DRIVERS_CRYPTO_CRYPTO_STM32_PRIV_H_
 #define ZEPHYR_DRIVERS_CRYPTO_CRYPTO_STM32_PRIV_H_
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
+#define STM32_CRYPTO_CONFIG_TYPEDEF            CRYP_InitTypeDef
```
Comment (erwango) [inline drivers/crypto/crypto_stm32_priv.h:12]: #define crypt_config_t

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/crypto/crypto_stm32.c (hunk 4 above)
Comment (lucasdietrich) [reply drivers/crypto/crypto_stm32.c]: Fixed

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for these changes, looks great.

Last couple of nitpicks, otherwise (and we'll be good to go):
- please update others L4 aes dts nodes
- at least one commit message needs rewording (the one with USE_HAL ...)
- and finally, we'll need to ensure this new code is compiled in CI, so there should be at least one matching in-tree board enabling this node.

**Comment by lucasdietrich** [comment]
I believe I have addressed most of the points.

However, to build the crypto sample on a board with a SoC that includes the ST, STM32L4-AES driver, I initially chose the `nucleo_l496zg`. However, it appears that this soc does not have an AES accelerator at all. According to the reference manual (RM0351), in the section "AES hardware accelerator (AES)," only the STM32L486/L4A6 include the AES accelerator.

The compilation actually fails because the header file `stm32cube/stm32l4xx/soc/stm32l496xx.h` from the STM32 HAL does not contain the "AES" constant required to enable the AES HAL.

I decided to check other SoCs:
- STM32L422xx (DS12470 Rev 6): AES accelerator is present.
- STM32L4P5 (DS12903 Rev 3): There is no clear mention of the AES device (surprisingly, there is a line for the AES device in the power consumption table).


Unless I am mistaken, I believe we should restrict the `st,stm32l4-aes` driver to STM32L462 and STM32L422xx SoCs.

As a result, there is currently no board in the Zephyr tree that uses an SoC with the STM32L4-AES accelerator. Furthermore, I am not aware of any development kit (DK) or known board that includes an STM32L422 or STM32L462.

**Comment by lucasdietrich** [comment]
> As a result, there is currently no board in the Zephyr tree that uses an SoC with the STM32L4-AES accelerator. Furthermore, I am not aware of any development kit (DK) or known board that includes an STM32L422 or STM32L462.

I just noticed that the STM32L4A6 SoC is present in the Zephyr tree. Should I add the AES node to the DTS for this SoC?

Morevoer, if a board for the [NUCLEO-L4A6ZG](https://www.st.com/en/evaluation-tools/nucleo-l4a6zg.html) could be added to the tree, it could serve as a candidate for "stm32l4-aes" testing in the CI.

File: samples/drivers/crypto/sample.yaml [hunk 10]
```diff
@@ -51,6 +51,22 @@ tests:
...
+    harness_config:
+      type: multi_line
+      regex:
+        - ".*: Cipher Sample"
+        - ".*: ECB mode ENCRYPT - Match"
+        - ".*: ECB mode DECRYPT - Match"
+        - ".*: CBC mode ENCRYPT - Match"
+        - ".*: CBC mode DECRYPT - Match"
+        - ".*: CTR mode ENCRYPT - Match"
+        - ".*: CTR mode DECRYPT - Match"
```
Comment (erwango) [inline samples/drivers/crypto/sample.yaml:69]: This is not required. What you just need to do is enable aes node directly in `nucleo_l4a6zg.dts` and then it will match filter of already existing test config `sample.drivers.crypto.stm32` thanks to `dt_compat_enabled("st,stm32-aes")`.

**Review by erwango — CHANGES_REQUESTED** [review]
One very last fix and we'll be good to go!

**Review by erwango — APPROVED** [review]
Approved, Thanks for your time!

**Comment by lucasdietrich** [comment]
You're welcome, just wanted to check if there's any update on merging the PR. Please let me know if I can help further.

**Review by gautierg-st — APPROVED** [review]
